### ISSUE TITLE
Code for analysis hyperspectral image

### ISSUE THREAD

**Issue by saikatcnu-alt** [saikatcnu-alt/SAIKAT-BISWAS#1]
import spectral
import numpy as np
import matplotlib.pyplot as plt
from spectral import open_image

# Use raw string (r"") to avoid escape character issues in Windows paths
hsi_image = open_image(r"C:/Users/Lorna/OneDrive/Desktop/python programme/DataToSaikat/2022_09_22_04_05_26(E, W)/west.hdr")

hsi_array = hsi_image.load()  # Load as NumPy array
hsi_array = hsi_array[:, :, :-1]  # Remove last band if needed

print(f"Original Shape: {hsi_image.shape}")
print(f"Modified Shape: {hsi_array.shape}")

#visualize the image
band = hsi_array[:, :, 50]
plt.imshow(band, cmap='grey')
plt.title('Band 70')
plt.show()

#Plot the spectral signature for specific pixel

pixel = hsi_array[100, 100, :]  # Extract spectral signature
pixel = np.squeeze(pixel)
plt.plot(pixel, marker='o')  # Add markers for better visibility
plt.title('Spectral Signature at (100, 100)')
plt.xlabel('Band Index')
plt.ylabel('Reflectance')
plt.grid(True)
plt.show()
#Normalizing Library

from sklearn.preprocessing import MinMaxScaler

#Normalizing Hyperspectral Data
def normalize_image(image):
    reshaped = image.reshape(-1, image.shape[-1])
    scaler = MinMaxScaler()
    normalized = scaler.fit_transform(reshaped).reshape(image.shape)
    return normalized
normalized_hsi =normalize_image(hsi_array)

#Dimensionality Reduction with PCA 
#(We Use PCA to reduce the redundency and efficient visualization and also for Noise Reduction
#PCA capture the essential information)

#PCA Library

from sklearn.decomposition import PCA

#Perform PCA

reshaped = hsi_array.reshape(-1,hsi_array.shape[-1])
pca = PCA(n_components=3)
reduced = pca.fit_transform(reshaped)

#Reshape Back
reduced_image = reduced.reshape(hsi_array.shape[0], hsi_array.shape[1], 3)
plt.imshow(reduced_image)
plt.title('PCA Reduced Image')
plt.show()

#####################################
from sklearn.cluster import KMeans

# --- Clustering ---
num_clusters = 4  # You can experiment (e.g., 3–6)
reshaped_reduced = reduced_image.reshape(-1, 3)
kmeans = KMeans(n_clusters=num_clusters, random_state=42)
labels = kmeans.fit_predict(reshaped_reduced)
clustered_image = labels.reshape(hsi_array.shape[0], hsi_array.shape[1])

# --- Display the clustered result ---
plt.imshow(clustered_image, cmap='tab10')
plt.title(f'K-Means Clustering ({num_clusters} Clusters)')
plt.colorbar(label='Cluster ID')
plt.show()
#####################################

# --- Band-wise Statistics ---
mean_spectrum = np.mean(hsi_array, axis=(0, 1))
std_spectrum = np.std(hsi_array, axis=(0, 1))

# --- Plot mean and std ---
plt.figure(figsize=(8, 4))
plt.plot(mean_spectrum, label='Mean Spectrum')
plt.fill_between(range(len(mean_spectrum)),
                 mean_spectrum - std_spectrum,
                 mean_spectrum + std_spectrum,
                 color='gray', alpha=0.3, label='±1 Std Dev')
plt.title('Mean Spectral Signature (All Pixels)')
plt.xlabel('Band Index')
plt.ylabel('Reflectance')
plt.legend()
plt.show()

#################################
import seaborn as sns

corr_matrix = np.corrcoef(hsi_array.reshape(-1, hsi_array.shape[-1]), rowvar=False)
plt.figure(figsize=(6, 5))
sns.heatmap(corr_matrix, cmap='coolwarm', cbar=True)
plt.title('Band Correlation Matrix')
plt.show()
